tool/abi.rb

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env ruby
+
+class ABI
+  IGNORED_PATTERNS = [
+    /include\/ruby\/internal\/abi.h\z/,
+    /include\/ruby-.+\/.+\/rb_mjit_min_header-.*\.h\z/,
+  ]
+
+  def initialize(srcdir)
+    @srcdir = srcdir
+    @output_file = "#{@srcdir}/include/ruby/internal/abi.h"
+  end
+
+  attr_reader :output_file
+
+  def version
+    begin
+      require "digest"
+    rescue LoadError
+      return 0
+    end
+    d = Digest::MD5.new
+
+    # Generate MD5 for every header file
+    header_files = Dir["#{@srcdir}/include/**/*.h"].reject { |f| skip_file?(f) }
+    header_files.sort!
+    md5s = header_files.map { |file| d.file(file) }
+
+    upper, lower = d.digest.unpack("Q>*")
+    upper ^ lower
+  end
+
+  private
+
+  def skip_file?(file)
+    IGNORED_PATTERNS.any? { |r| r =~ file }
+  end
+end
+
+# Run only if file was directly called
+if __FILE__ == $0
+  require "optparse"
+  require_relative "lib/colorize"
+
+  OptionParser.new do |opts|
+    opts.on("--srcdir=PATH", "use PATH as source directory") do |srcdir|
+      @srcdir = srcdir
+    end
+  end.parse!
+
+  abi = ABI.new(@srcdir)
+  color = Colorize.new(color)
+  updated = color.fail("updated")
+
+  abi_h = <<-C
+#ifndef RUBY_ABI_H
+#define RUBY_ABI_H
+
+// DO NOT MODIFY THIS FILE DIRECTLY
+// This file was automatically generated by tool/abi.rb
+
+RUBY_FUNC_EXPORTED unsigned long long __attribute__((weak))
```

Comment:
`__attribute__((weak))` might not exist.

Reply:
Maybe you just want to make it `static inline`.

Reply:
Ah, I just read `dln.c` part.  I know you intentionally want a weak function.  `static inline` is inappropriate then.  Sorry.

Reply:
But it is true that there are compilers without that attribute.  Consider checking `HAVE_FUNC_WEAK` instead.

Reply:
Thank you! I fixed this in the most recent push.

Reply:
Thanks! Looks good to me now.